### DISCUSSION TITLE
Is it possible in C# to specify that a generic type can have any number of type arguments?

### DISCUSSION THREAD

**Discussion by mertmtn** [dotnet/csharplang#6481] [Language Ideas]
I have a sample code like below which created using generic classes. It is just proposed sample code. I want to send more than one mapping profiler like generic type params <T[]>

```
using AutoMapper; 

namespace Northwind.Core.Entities.Mapping.AutoMapper
{ 
    public class AutoMapperMapper<TProfile[]> : IMapper where TProfile[] : Profile[], new()
    {
        public TDestination Map<TSource, TDestination>(TSource source)
        {
            var mapper = new MapperConfiguration(config =>
            {
                foreach (var item in TProfile[])
                {
                    config.AddProfile(new item); 
                }
            }).CreateMapper();

            return mapper.Map<TDestination>(source);
        }
    } 
} 
```

My purpose is using params with dependency injection. 

`services.AddSingleton<IMapper, AutoMapperMapper<CategoryProfile, ProductProfile>();`

Is it possible in C# to specify that a generic type can have any number of type arguments? I guess there's no implementation like params to specify multiple type arguments.

By the way AutoMapper is library used for this code to map objects.  

I think it can be required for future.

**Comment by CyrusNajmabadi** [comment]
There is no way currently. General convention is to provide up through 4-8 overloads, which tends to hit the 99%+ in practice.

**Comment by TahirAhmadov** [comment]
Actually I think this calls for an interface, `IProfile`:
```cs
using AutoMapper; 

namespace Northwind.Core.Entities.Mapping.AutoMapper
{ 
    public class AutoMapperMapper : IMapper
    {
        public AutoMapperMapper(IProfile[] profiles) => this._profiles = profiles;
        public TDestination Map<TSource, TDestination>(TSource source)
        {
            var mapper = new MapperConfiguration(config =>
            {
                foreach (var profile in this._profiles)
                {
                    config.AddProfile(profile); 
                }
            }).CreateMapper();

            return mapper.Map<TDestination>(source);
        }
    } 
} 
```

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/2330
